amethyst_locale/src/lib.rs

```diff
@@ -49,6 +50,9 @@ pub struct Locale {
     pub context: MessageContext<'static>,
 }
 
+unsafe impl Sync for Locale {}
+unsafe impl Send for Locale {}
```

Comment:
Please clarify the safety of this operation.

Reply:
I need to look into MessageContext being thread safe, so I was going to open an issue for that.

If `Send` and `Sync` are not implemented, you get this lovely error:
```rust
error[E0277]: `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be shared between threads safely
  --> amethyst_locale/src/lib.rs:24:6
   |
24 | impl SimpleFormat<Locale> for LocaleFormat {
   |      ^^^^^^^^^^^^^^^^^^^^ `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be shared between threads safely
   |
   = help: the trait `std::marker::Sync` is not implemented for `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `std::boxed::Box<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `fluent::context::Entry<'static>`
   = note: required because it appears within the type `(std::string::String, fluent::context::Entry<'static>)`
   = note: required because it appears within the type `std::marker::PhantomData<(std::string::String, fluent::context::Entry<'static>)>`
   = note: required because it appears within the type `std::collections::hash::table::RawTable<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `std::collections::HashMap<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `fluent::MessageContext<'static>`
   = note: required because it appears within the type `Locale`

error[E0277]: the trait bound `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static: std::marker::Send` is not satisfied
  --> amethyst_locale/src/lib.rs:24:6
   |
24 | impl SimpleFormat<Locale> for LocaleFormat {
   |      ^^^^^^^^^^^^^^^^^^^^ `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `std::boxed::Box<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `fluent::context::Entry<'static>`
   = note: required because it appears within the type `(std::string::String, fluent::context::Entry<'static>)`
   = note: required because it appears within the type `std::marker::PhantomData<(std::string::String, fluent::context::Entry<'static>)>`
   = note: required because it appears within the type `std::collections::hash::table::RawTable<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `std::collections::HashMap<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `fluent::MessageContext<'static>`
   = note: required because it appears within the type `Locale`

error[E0277]: `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be shared between threads safely
  --> amethyst_locale/src/lib.rs:53:6
   |
53 | impl Asset for Locale {
   |      ^^^^^ `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be shared between threads safely
   |
   = help: the trait `std::marker::Sync` is not implemented for `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `std::boxed::Box<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `fluent::context::Entry<'static>`
   = note: required because it appears within the type `(std::string::String, fluent::context::Entry<'static>)`
   = note: required because it appears within the type `std::marker::PhantomData<(std::string::String, fluent::context::Entry<'static>)>`
   = note: required because it appears within the type `std::collections::hash::table::RawTable<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `std::collections::HashMap<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `fluent::MessageContext<'static>`
   = note: required because it appears within the type `Locale`

error[E0277]: the trait bound `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static: std::marker::Send` is not satisfied
  --> amethyst_locale/src/lib.rs:53:6
   |
53 | impl Asset for Locale {
   |      ^^^^^ `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `std::boxed::Box<for<'r, 's> std::ops::Fn(&'r [std::option::Option<fluent::types::FluentValue>], &'s std::collections::HashMap<std::string::String, fluent::types::FluentValue>) -> std::option::Option<fluent::types::FluentValue> + 'static>`
   = note: required because it appears within the type `fluent::context::Entry<'static>`
   = note: required because it appears within the type `(std::string::String, fluent::context::Entry<'static>)`
   = note: required because it appears within the type `std::marker::PhantomData<(std::string::String, fluent::context::Entry<'static>)>`
   = note: required because it appears within the type `std::collections::hash::table::RawTable<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `std::collections::HashMap<std::string::String, fluent::context::Entry<'static>>`
   = note: required because it appears within the type `fluent::MessageContext<'static>`
   = note: required because it appears within the type `Locale`

error: aborting due to 4 previous errors
```

Reply:
Okay but right now if you can't prove it's actually thread safe we can't just implement it unsafely.
Is there really no alternative?

Reply:
I'll wrap it in a `Mutex`, and create an issue to check if it's safe to remove the mutex and use the raw `MessateContext` instead.

Reply:
Well if it's not Send, the mutex might not be enough.
I'd assume this is not thread safe only because they use the thread-unsafe flavors of common types, it might be easy to fix with a PR to them.

Reply:
I'll revert the `fluent` upgrade then

Reply:
If you want to spend some additional time trying to fix that on their side it would be really appreciated, but `fluent` isn't the most important crate to keep up to date so you should not worry too much about that.